consensus/istanbul/types.go

```diff
@@ -343,6 +425,65 @@ type Message struct {
 	Msg       []byte
 	Address   common.Address // The sender address
 	Signature []byte         // Signature of the Message using the private key associated with the "Address" field
+
+	// The below fields are not serializable since they are private, they are
+	// set when calling Message.FromPayload, only one will be set in any
+	// instance, which is set depends on the Message.Code.
+	committedSubject    *CommittedSubject
+	prePrepare          *Preprepare
+	prepare             *Subject
+	roundChange         *RoundChange
+	queryEnode          *QueryEnodeData
+	forwardMessage      *ForwardMessage
+	enodeCertificate    *EnodeCertificate
+	versionCertificates []*VersionCertificate
+	valEnodeShareData   *ValEnodesShareData
+}
+
+// NewMessage constructs a message with the innerMessage instance and the
+// sender address, it will panic if innerMessage is not a recognised inner
+// message type.
+func NewMessage(innerMessage interface{}, sender common.Address) *Message {
+	message := &Message{
+		Address: sender,
+	}
+	switch t := innerMessage.(type) {
+	case *Preprepare:
+		message.Code = MsgPreprepare
+		message.prePrepare = t
+	case *Subject:
+		message.Code = MsgPrepare
+		message.prepare = t
+	case *CommittedSubject:
+		message.Code = MsgCommit
+		message.committedSubject = t
+	case *RoundChange:
+		message.Code = MsgRoundChange
+		message.roundChange = t
+	case *QueryEnodeData:
+		message.Code = QueryEnodeMsg
+		message.queryEnode = t
+	case *ForwardMessage:
+		message.Code = FwdMsg
+		message.forwardMessage = t
+	case *EnodeCertificate:
+		message.Code = EnodeCertificateMsg
+		message.enodeCertificate = t
+	case []*VersionCertificate:
+		message.Code = VersionCertificatesMsg
+		message.versionCertificates = t
+	case *ValEnodesShareData:
+		message.Code = ValEnodesShareMsg
+		message.valEnodeShareData = t
+	default:
+		panic(fmt.Sprintf("attempt to construct Message unrecognised inner message type %T", t))
```

Comment:
Is there any way to make this a compile-time error rather than a panic?

Reply:
I can't see any neat way to convert this into a compile time error, can you?

Reply:
The way to make this a compile time error would to have a `func NewTypeMessage(msg Type, sender common.Address) *Msg ` for each type. It's probably a similar amount of total code, but way more functions.

Reply:
@trianglesphere Thanks, I had considered that, but I was thinking it would be quite a lot more code. Thinking about it more I'm not so sure that it would be. I could open a PR on this to see what that change looks like.